data-prepper-plugins/saas-source-plugins/saas-crawler/src/main/java/org/opensearch/dataprepper/plugins/source/saas/crawler/coordination/PartitionFactory.java

```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.dataprepper.plugins.source.saas.crawler.coordination;
+
+import org.opensearch.dataprepper.model.source.coordinator.SourcePartitionStoreItem;
+import org.opensearch.dataprepper.model.source.coordinator.enhanced.EnhancedSourcePartition;
+import org.opensearch.dataprepper.plugins.source.saas.crawler.coordination.partition.LeaderPartition;
+import org.opensearch.dataprepper.plugins.source.saas.crawler.coordination.partition.SaasSourcePartition;
+
+import java.util.function.Function;
+
+/**
+ * Special partition factory just SAAS connector source.
+ */
+public class PartitionFactory implements Function<SourcePartitionStoreItem, EnhancedSourcePartition> {
+
+
+    @Override
+    public EnhancedSourcePartition apply(SourcePartitionStoreItem partitionStoreItem) {
+        String sourceIdentifier = partitionStoreItem.getSourceIdentifier();
+        String partitionType = sourceIdentifier.substring(sourceIdentifier.lastIndexOf('|') + 1);
+
+         if (LeaderPartition.PARTITION_TYPE.equals(partitionType)) {
+            return new LeaderPartition(partitionStoreItem);
+        } else if (SaasSourcePartition.PARTITION_TYPE.equals(partitionType)) {
+            return new SaasSourcePartition(partitionStoreItem);
+        } else {
+            // Unable to acquire other partitions.
+            // Probably we will introduce Global state in the future but for now, we don't expect to reach here.
+             throw new RuntimeException("Unable to acquire other partitions. " +
```

Comment:
Maybe print out the partitionType here if we get to this point.

Reply:
Added to the exception message.